### PR TITLE
Honor model dtype in `load_checkpoint`

### PR DESCRIPTION
**Pull request by sgugger**
This PR fixes a standing bug where we have a different behavior than PyTorch. In torch, loading a `state_dict` inside a model will never change the model's dtype:

```py
import torch

model = torch.nn.Linear(5, 6)
model_in_half = model.half()
state_dict_in_half = model.state_dict()

model = torch.nn.Linear(5, 6).to("meta")
model.load_state_dict(state_dict_in_half)
model.weight.dtype
# return torch.float32
```

Currently in Accelerate, `load_checkpoint` does the opposite and when loading a model, it converts it to the dtype of the state dict. This PR addresses that.

This PR only contains the fix for now, we have to discuss how to maybe maintain backward compatibility (even if this is a bug fix), because `diffusers` might be relying on this behavior, cc @patrickvonplaten

### CODE REVIEW

**Review by muellerzr — APPROVED** [review]
Thanks, that makes sense. I don't particularly think there is a "harm" in silently pushing it out (i.e. don't advertise the bad behavior but let it still pass) in this particular case. If we do care about phasing that out perhaps leave it for a 1.0.0? (Similar to some optimizer bits we have)

**Comment by patrickvonplaten** [comment]
Actually before merging, should be maybe not better handle this in `set_module_tensor_to_device` ? E.g. add a `dtype` argument to the function there? This would be easier for `diffusers` to by in line with `accelerate` I think - see: https://github.com/huggingface/diffusers/blob/727434c206f6c22b746e460293035a1324f0bc13/src/diffusers/modeling_utils.py#L491

File: src/accelerate/utils/modeling.py [hunk 1]
```diff
@@ -680,8 +694,7 @@ def load_checkpoint_in_model(
         else:
             for param_name, param in checkpoint.items():
                 module_name = param_name
-                if dtype is not None and not str(param.dtype).startswith(("torch.uint", "torch.int", "torch.bool")):
```
Comment (sgugger) [inline src/accelerate/utils/modeling.py:683]: This is moved to `set_module_tensor_to_device`.

**Review by patrickvonplaten — APPROVED** [review]
Thanks a lot for adapting!

**Pull request merged**